src/codal_port/microbit_microphone.c

```diff
@@ -147,13 +148,89 @@ STATIC mp_obj_t microbit_microphone_get_events(mp_obj_t self_in) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(microbit_microphone_get_events_obj, microbit_microphone_get_events);
 
+STATIC mp_obj_t microbit_microphone_record(mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_duration, ARG_rate, };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_duration, MP_ARG_REQUIRED | MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_rate, MP_ARG_INT, {.u_int = 7812} },
+    };
+
+    // Parse the args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Create the AudioFrame to record into.
+    size_t size = args[ARG_duration].u_int * args[ARG_rate].u_int / 1000;
+    microbit_audio_frame_obj_t *audio_frame = microbit_audio_frame_make_new(size, args[ARG_rate].u_int);
+
+    // Start the recording.
+    microbit_hal_microphone_start_recording(audio_frame->data, audio_frame->alloc_size, &audio_frame->used_size, args[ARG_rate].u_int);
+
+    // Wait for the recording to finish.
+    while (microbit_hal_microphone_is_recording()) {
+        mp_handle_pending(true);
+        microbit_hal_idle();
+    }
+
+    // Return the new AudioFrame.
+    return MP_OBJ_FROM_PTR(audio_frame);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_KW(microbit_microphone_record_obj, 1, microbit_microphone_record);
+
+STATIC mp_obj_t microbit_microphone_record_into(mp_uint_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_buffer, ARG_rate, ARG_wait, };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_buffer, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_rate, MP_ARG_INT, {.u_int = 0} },
+        { MP_QSTR_wait, MP_ARG_KW_ONLY | MP_ARG_BOOL, {.u_bool = true} },
+    };
+
+    // Parse the args.
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    // Check that the buffer is an AudioFrame instance.
+    if (!mp_obj_is_type(args[ARG_buffer].u_obj, &microbit_audio_frame_type)) {
+        mp_raise_TypeError(MP_ERROR_TEXT("expecting an AudioFrame"));
+    }
+    microbit_audio_frame_obj_t *audio_frame = MP_OBJ_TO_PTR(args[ARG_buffer].u_obj);
+
+    // Set the rate of the AudioFrame, if specified.
+    if (args[ARG_rate].u_int > 0) {
+        audio_frame->rate = args[ARG_rate].u_int;
+    }
+
+    // Start the recording.
+    microbit_hal_microphone_start_recording(audio_frame->data, audio_frame->alloc_size, &audio_frame->used_size, audio_frame->rate);
+
+    return mp_const_none;
```

Comment:
I think this is missing the wait flag implementation.

For example, with this bit of code, the display shows "R" inmediately:
```python
my_recording = audio.AudioFrame(duration=3000)
microphone.record_into(my_recording, wait=True)
display.show("R")
```

Reply:
`wait` is now implemented.